### ISSUE TITLE
Fatal error while rendering plots when `base::ifelse()` is masked by a stricter R function

### ISSUE THREAD

**Issue by mattsams89** [quarto-dev/quarto-cli#2454]
### Bug description

I wasn't completely sure if this was better suited for the cli repo or the R repo, but here we are. The `tidytable` package recently introduced dotless versions of functions in its dev version (e.g. `mutate.() -> mutate()`), with one of them being its own version of `base::ifelse()` that ultimately calls `data.table::fifelse()` under the hood. It seems the stricter type checking of `fifelse` isn't playing nicely with a function that's calling `knitr`'s plot hook functions. The below fails on both html and pdf rendering:

````
---
format: html
---

```{r}
data("mtcars")

# devtools::install_github("markfairbanks/tidytable")

library(tidytable)

plot(mtcars$mpg, mtcars$cyl)
```
````

```
processing file: plot_example.qmd
  |.......................                                               |  33%
  ordinary text without R code

  |...............................................                       |  67%
label: unnamed-chunk-1
Quitting from lines 6-13 (plot_example.qmd) 
Error in fifelse(condition, args$true, args$false, args$missing) : 
  'yes' is of type NULL but 'no' is of type character. Please make sure that both arguments have the same type.
Calls: .main ... run_hook_plot -> hook -> ifelse -> if_else -> fifelse

Execution halted
```

I haven't had the time to read through the rmd-related code super closely, but I *think* it may be coming from [line 162](quarto-dev/quarto-cli/blob/main/src/resources/rmd/execute.R#L162) of `execute.R` since that's the only `ifelse` I've come across so far that seems like it could return two different types.

For posterity and completeness' sake, both PCs I've tested this on are running Windows 10 with the 1.1.168 cli. Any dev version of `tidytable` after [this commit](https://www.github.com/markfairbanks/tidytable/commit/83b826060a126c1974d18c4646b6739fdbfdbedc) will produce the error. Finally, the error doesn't occur if knitting an .rmd with the same code as above.

```
quarto help

 Usage:   quarto
 Version: 1.1.168

Description:

Quarto CLI
```

```
packageVersion("knitr")

1.39
```

In the meantime, I've been calling `conflicted::conflict_prefer("ifelse", "base")` to get around the issue.

### Checklist

- [X] Please include a minimal, fully reproducible example in a single .qmd file? Please provide the whole file rather than the snippet you believe is causing the issue.
- [X] Please [format your issue](https://quarto.org/bug-reports.html#formatting-make-githubs-markdown-work-for-us) so it is easier for us to read the bug report.
- [x] Please document the RStudio IDE version you're running (if applicable), by providing the value displayed in the "About RStudio" main menu dialog?
- [X] Please document the operating system you're running. If on Linux, please provide the specific distribution.

**Comment by cscheid** [comment]
Thanks for the report. To make sure I understand this correctly: `tidytable` exports a function that overrides a function from `base` and has different behavior from it? If so, isn't that a problem with `tidytable`'s `ifelse`?

**Comment by mattsams89** [comment]
Right, it enforces type/class consistency requirements like you would see in `data.table::fifelse()` or `dplyr::if_else()`. Personally, I think this is a pretty great example of why these other functions enforce type consistency in the first place—to avoid unexpected behavior that tends to be a head scratcher on the diagnostic side of things. And sure, the current functionality of the base function is unlikely to change, but it makes far more sense in my mind to explicitly handle the true and false instead of relying on the base function continuing to allow different types and classes in its true and false. 

Talking to tidytable’s dev, seems like he’s going to disable the exported function for now , but may be a worthwhile forward-thinking adjustment to consider regardless.

**Comment by cderv** [comment]
@cscheid I think we can have other issues like that as our current implementation doesn't have yet the security of namespacing function. I don't think it is only a third party issue that a package use same name as another package - not ideal sure, but I think we should be robust to that. 

I mean a quick fix for this would be 
````diff
-ifelse(!is.null(libDir), libDir, files_dir),  
+base::ifelse(!is.null(libDir), libDir, files_dir), 
```` 

but we should maybe do more general to prevent any other issue. Let's discuss on Slack.

**Comment by mattsams89** [comment]
Namespacing it is definitely the obvious solution my brain was too frazzled to think about yesterday.

**Comment by cderv** [comment]
I am trying to look if we have a way to execute our R code in a way that it will be isolated from such issue. Namespacing our function will be the solution otherwise. 

@cscheid I suggest we reopen and I can be in charge of this. for v1.2 ?

**Comment by cscheid** [comment]
Sure!

**Comment by cderv** [comment]
I understand this was resolved upstream, and we did not encountered any others issues like this since. 

I still believe we are not completely safe as currently we are sourcing script and runnning function before `rmarkdown::render()` but also applying functions during knitting (with hooks) and after `rmarkdown::render()`. issue could be in those two steps as .qmd R code content could masked some of our script internal function because we have no lock namespace (as you do when writing functions in package). 

> Namespacing our function will be the solution otherwise.

I looked closer at functions we are using and it mostly base function which are un-namespace. It feels quite unusual to namespace `base::fun()`. Usually, R code from packages and user code is not expected to touch base R function, but it happens. 
I have a [branch](https://github.com/quarto-dev/quarto-cli/compare/main...cderv:quarto-cli:r-namespace-files) which namespace every function, but it feels a bit unreadable. So not sure that worth merging yet. 

I'll keep that around, and ultimately I think it would useful to find a way to isolate the execution that can cause issues. I always have in mind solutions like [**box**](https://cran.r-project.org/web/packages/box/index.html) 📦 . 

Anyway, I'll close this for now.
